### ISSUE TITLE
Input placeholder with shape [1] of saved_model.pb accepts scaler input but throws error in later graph.

### ISSUE THREAD

**Issue by Mickky666** [tensorflow/serving#1436]
Env:
* Mac OS
* python 2.7
    * tensorflow 1.13.1 
    * tensorflow-serving-api 1.20.1
    * grpcio 1.12.0
* docker 19.03.1
* serving docker image: `tensorflow/serving:1.13.0`

Description:
I'm using estimator API to build model. Inside serving input function, my model has an int64 input placeholder with shape [1]. And this placeholder is later concatenated with another tensor. Then I trained the model and saved it to `*.pb` graph. However during serving time I mistakenly feed a scaler value to this placeholder:
```
dtype: DT_INT64
tensor_shape {
}
int64_val: 182
```
As you can see, there is no dimension info in this proto. It didn't immediately throw an error and it seems that the placeholder with shape [1] is compatible with it. However later during concatenating op I got the error:
```
*** _Rendezvous: <_Rendezvous of RPC that terminated with:
        status = StatusCode.INVALID_ARGUMENT
        details = "ConcatOp : Expected concatenating dimensions in the range [0, 0), but got 0
         [[{{node concat_77}}]]"
        debug_error_string = "{"created":"@1567706789.859604000","description":"Error received from peer ipv6:[::1]:9000","file":"src/core/lib/surface/call.cc","file_line":1041,"grpc_message":"ConcatOp : Expected concatenating dimensions in the range [0, 0), but got 0\n\t [[{{node concat_77}}]]","grpc_status":3}"
```
Later I realized I did something wrong in the dimension so I add the shape info to my input proto. The error is gone. The correct proto looks like:
```
dtype: DT_INT64
tensor_shape {
  dim {
    size: 1
  }
}
int64_val: 182
```
In one word, a `saved_model.pb` with input placeholder of shape [1] accepts a scaler input and not throwing any error. However later in the concat operation, since we are doing `tf.concat([scaler, tf.constant([1])], axis=0)`, it throws an error since the scaler has 0 dimension.

Code to reproduce the error:
```
import tensorflow as tf
import numpy as np
import os
import grpc
from tensorflow_serving.apis import predict_pb2
from tensorflow_serving.apis import prediction_service_pb2_grpc
from tensorflow.python.estimator.inputs import numpy_io


x = {"my_feature": np.random.randint(0, 20, size=(20, 2))}
with tf.Session() as session:
    train_input_fn = numpy_io.numpy_input_fn(x, batch_size=1, shuffle=False, num_epochs=20)


def serve_input_fn():
    receiver_tensors = {}
    serving_features = {}
    receiver_tensors['my_input'] = tf.placeholder(shape=[1], dtype=tf.int64)
    serving_features['my_feature'] = tf.concat([receiver_tensors['my_input'], [1]], axis=0)
    return tf.estimator.export.ServingInputReceiver(serving_features, receiver_tensors)


def model_fn(features, labels, mode):
    var = tf.get_variable("dummy_weights", shape=(2,), dtype=tf.float32)
    output = tf.reduce_sum(var * tf.cast(features['my_feature'], dtype=tf.float32))
    loss = (output - 10) ** 2
    train_op = tf.train.AdamOptimizer(1e-3).minimize(loss)
    if mode == tf.estimator.ModeKeys.TRAIN:
        return tf.estimator.EstimatorSpec(mode=mode, loss=loss, train_op=train_op)
    if mode == tf.estimator.ModeKeys.PREDICT:
        return tf.estimator.EstimatorSpec(mode=mode, predictions={'output': output})


estimator = tf.estimator.Estimator(
    model_fn=model_fn,
    model_dir="/tmp/model_test/",
)
estimator.train(train_input_fn, steps=10)
# create any estimater
estimator.export_saved_model("/tmp/my_model", serve_input_fn)


def serve():
    docker_kill_cmd = r"docker kill `docker ps -q`"
    docker_remove_cmd = r"docker rm `docker ps -q -a`"
    combined_cmd = docker_kill_cmd + " && " + docker_remove_cmd
    os.system(combined_cmd)
    docker_cmd = r"""
                                    docker run -d -p {port}:8500 
                                    --mount type=bind,source={model_dir},target=/models/model/ 
                                    -e MODEL_NAME={model_name} -t tensorflow/serving:1.13.0
                                    --tensorflow_intra_op_parallelism=12
                                    --tensorflow_inter_op_parallelism=12
                                """.format(model_dir="/tmp/my_model/", model_name="model", port=9000).replace('\n', '')
    ret_status = os.system(docker_cmd)
    if ret_status != 0:
        raise Exception("system call fail: {}".format(docker_cmd))
    channel = grpc.insecure_channel("localhost:9000")
    stub = prediction_service_pb2_grpc.PredictionServiceStub(channel)
    return stub


# setup serving client
serving_stub = serve()

# build request proto
request = predict_pb2.PredictRequest()
request.model_spec.name = "model"
request.model_spec.signature_name = "serving_default"
request.inputs['my_input'].CopyFrom(tf.contrib.util.make_tensor_proto(10, dtype=tf.int64))
output = serving_stub.Predict(request, 10)
```

**Comment by Mickky666** [comment]
CC @ravikyram

**Comment by gowthamkpr** [comment]
Can you please provide the complete code (github gist) for us to reproduce the issue. Thanks!

**Comment by Mickky666** [comment]
@gowthamkpr 
Sure, the gist is here: https://gist.github.com/Mickky666/59e8891bccbe93480b9a5a735e514b97
Also attached the python environment here:
https://gist.github.com/Mickky666/1922b48cc340f37950906f5d9b3c99ff

**Comment by gowthamkpr** [comment]
@Mickky666 Can you please provide the signature def that you used while saving the model. thanks!

**Comment by Mickky666** [comment]
@gowthamkpr Sure, the model signature of the above example code is:
```
signature_def['serving_default']:
  The given SavedModel SignatureDef contains the following input(s):
    inputs['my_input'] tensor_info:
        dtype: DT_INT64
        shape: (1)
        name: Placeholder:0
  The given SavedModel SignatureDef contains the following output(s):
    outputs['output'] tensor_info:
        dtype: DT_FLOAT
        shape: ()
        name: Sum:0
  Method name is: tensorflow/serving/predict
```
The input `my_input` has shape (1), however it can accept a raw scaler input like:
```
tf.contrib.util.make_tensor_proto(10, dtype=tf.int64)
```

**Comment by gowthamkpr** [comment]
@Mickky666 Modelserver does not do any input shape verification while handling a predict request.

typically the graph processing imposes certain restrictions on the tensors. and invalid shapes usually fail during graph execution as seen in the issue.

some graphs are flexible on dimensions they accept (like our half_plus_two model, that does 0.5*x+2 for tensors of varied dimensions). so it is arguable if we should do any verification. then there is this implicit batch size as the first dimension, whose value is not known in advance.

**Comment by gowthamkpr** [comment]
@Mickky666 Closing this issue as it has been resolved. Please add additional comments and we can open the issue again. Thanks!
